### PR TITLE
Support incremental testing of dbt models through failure thresholds

### PR DESCRIPTION
**Pull request by jeancochrane**
We know there are some existing integrity issues with our data that the data catalog is designed to help us address, but we don't want our dbt tests to unnecessarily alert us about these known issues while we're developing the catalog. To handle this problem, this PR adds some simple thresholds to the tests to ensure that they only fail when the number of test failures exceeds the number of expected failures.

The numbers used in these thresholds may not accurately reflect underlying integrity issues, and instead may simply represent problems with the tests themselves; we haven't thoroughly audited our tests yet so we can't rule that out. But I'm considering that effort out of scope for now, and will instead be a part of https://github.com/ccao-data/data-architecture/issues/27.

Adding these failure thresholds was significantly easier than I thought it would be (#32) since it turns out test severity thresholds [have native support in dbt](https://docs.getdbt.com/reference/resource-configs/severity). Nice! In case we want to support running these tests with the option to ignore the thresholds, we can do so with the following incantation:

```
$ dbt --warn-error test
```

The PR also makes two unrelated changes that helped me debug failing tests:

* Fix a bug that was preventing the build of the `default.vw_pin_sale` model due to a missing column in a subquery
* Add simple model namespacing to tests to work around the fact that dbt does not 

Closes https://github.com/ccao-data/data-architecture/issues/32.

### CODE REVIEW

File: aws-athena/views/default-vw_pin_sale.sql [hunk 1]
```diff
@@ -185,7 +185,8 @@ mydec_sales AS (
             ) > 0 AS sale_filter_ptax_flag,
             COUNT() OVER (
                 PARTITION BY line_1_primary_pin, line_4_instrument_date
-            ) AS num_cards_sale
+            ) AS num_cards_sale,
+            year_of_sale
```
Comment (jeancochrane) [inline aws-athena/views/default-vw_pin_sale.sql:189]: This missing column was causing `dbt run` to fail to build this model:

```
18:53:31  Runtime Error in model vw_pin_sale_test (models/default/vw_pin_sale_test.sql)
18:53:31    line 212:18: Column 'mydec_sales.year_of_sale' cannot be resolved
```

I had forgotten that I had made this adjustment prior to factoring out shared logic between the Athena and dbt views in https://github.com/ccao-data/data-architecture/pull/41.

Perhaps this change should also be pulled into `master`? It seems like the Athena version of the view correctly selects this column, so it must just be a discrepancy between the production version of the view and the version stored in this file. (Another nice category of problem that dbt will eventually resolve for us.)

File: dbt/models/default/schema.yml [hunk 2]
```diff
@@ -56,36 +63,46 @@ models:
...
           # char_renovation can be null, which we count here as falsey
           expression: >-
             char_recent_renovation = (
               case when char_renovation = '1' then true else false end
             )
+          config:
+            # This error threshold represents an upper bound, since the
+            # underlying query that generates `char_recent_renovation`
+            # is currently not deterministic. When we fix the underlying
+            # view to make that column deterministic, we should lower this.
```
Comment (jeancochrane) [inline dbt/models/default/schema.yml:91]: Very strange! I think this has to do with the ordered window functions that we use to generate `vw_card_res_char.char_recent_renovation`:

https://github.com/ccao-data/data-architecture/blob/73cdfec06aa2730d41a06558bcf469b4a3f9c124/dbt/models/default/vw_card_res_char_test.sql#L90-L124

Billy's looking into it!

File: dbt/models/default/schema.yml [hunk 3]
```diff
@@ -7,7 +7,7 @@ models:
     tests:
       # Unique by 14-digit PIN and year
       - unique_combination_of_columns:
-          name: unique_by_14_digit_pin_and_year
+          name: vw_pin_history_unique_by_14_digit_pin_and_year
```
Comment (jeancochrane) [inline dbt/models/default/schema.yml:10]: An example of what test output looks like before adding this namespacing, to motivate the change:

```
19:53:13  1 of 19 START test unique_by_14_digit_pin_and_year ................... [RUN]
```

There are a bunch of tests named `unique_by_14_digit_pin_and_year` in particular.

**Comment by jeancochrane** [comment]
Requesting both @dfsnow and @wrridgeway on this for the sake of keeping everyone up to date on how we're handling failure thresholds, but I only really need one thorough review!

File: aws-athena/views/default-vw_pin_sale.sql (hunk 1 above)
Comment (dfsnow) [reply aws-athena/views/default-vw_pin_sale.sql:189]: Can you cherry-pick this commit and open a PR to master? This seems like a mistake

File: dbt/models/default/schema.yml (hunk 2 above)
Comment (dfsnow) [reply dbt/models/default/schema.yml:91]: Seemingly fixed by #47.

Comment (jeancochrane) [reply dbt/models/default/schema.yml]: Yup, that fixes it -- resolving this now!

File: aws-athena/views/default-vw_pin_sale.sql (hunk 1 above)
Comment (jeancochrane) [reply aws-athena/views/default-vw_pin_sale.sql:189]: Can do! [PR up here](https://github.com/ccao-data/data-architecture/pull/48), I'll request your review once tests pass.

File: dbt/models/default/schema.yml [hunk 4]
```diff
@@ -97,17 +114,21 @@ models:
             - mail_address_state
             - mail_address_zipcode_1
             - mail_address_zipcode_2
+          config:
+            error_if: ">879261"
```
Comment (dfsnow) [inline dbt/models/default/schema.yml:114]: **issue (non-blocking)**: This high error rate is a result of using `CONCAT_WS` to construct `prop_address_full` and `mail_address_full`. Any address missing a unit number or description will have trailing whitespace. See https://github.com/ccao-data/data-architecture/blob/36a42a7d304577fd77440ff1d807905d5bae1a75/aws-athena/views/default-vw_pin_address.sql#L20C12-L20C21.

@wrridgeway Can you make a new PR to master wrapping `CONCAT_WS` in `TRIM()` for these address fields.

@jeancochrane We can keep this PR moving but we'll need to remember to update this threshold once Billy's PR is merged.

File: dbt/models/default/schema.yml (hunk 3 above)
Comment (dfsnow) [reply dbt/models/default/schema.yml:10]: Love this, good thinking!

**question (non-blocking)**: Is there no way to define the namespace in the test definition? If not, this totally works.

**Review by dfsnow — APPROVED** [review]

File: aws-athena/views/default-vw_pin_sale.sql (hunk 1 above)
Comment (jeancochrane) [reply aws-athena/views/default-vw_pin_sale.sql:189]: Merged to `master` 👍🏻

**Pull request merged**